### PR TITLE
expression: add builtin func bit_length() for gbk

### LINKED ISSUE #29460: Support GBK for builtin function `bit_length`

### PR DESCRIPTION
**Pull request by chacha923**
### What problem does this PR solve?

Issue Number: close #29460

Problem Summary: add builtin func bit_length() for gbk

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: expression/builtin_string.go [hunk 1]
```diff
@@ -2358,7 +2358,13 @@ func (b *builtinBitLengthSig) evalInt(row chunk.Row) (int64, bool, error) {
 	if isNull || err != nil {
 		return 0, isNull, err
 	}
-
+	argTp := b.args[0].GetType()
+	if !types.IsBinaryStr(argTp) {
+		dBytes, err := charset.NewEncoding(argTp.Charset).Encode(nil, hack.Slice(val))
+		if err == nil {
```
Comment (Defined2014) [inline expression/builtin_string.go:2364]: perfer return error when `err != nil`

File: expression/builtin_string.go [hunk 2]
```diff
@@ -2358,7 +2358,13 @@ func (b *builtinBitLengthSig) evalInt(row chunk.Row) (int64, bool, error) {
 	if isNull || err != nil {
 		return 0, isNull, err
 	}
-
+	argTp := b.args[0].GetType()
+	if !types.IsBinaryStr(argTp) {
```
Comment (Defined2014) [inline expression/builtin_string.go:2362]: IMO, it doesn't need anymore.

File: expression/builtin_string_test.go [hunk 3]
```diff
@@ -1417,6 +1417,28 @@ func TestBitLength(t *testing.T) {
...
 	_, err := funcs[ast.BitLength].getFunction(ctx, []Expression{NewZero()})
 	require.NoError(t, err)
+
+	// Test GBK String
+	tbl := []struct {
+		input  string
+		chs    string
+		result int64
+	}{
+		{"abc", "gbk", 24},
```
Comment (Defined2014) [inline expression/builtin_string_test.go:1427]: Could you add the test cases into previous tests? like add `chs` in `cases`

**Comment by chacha923** [comment]
PTAL @Defined2014

**Comment by Defined2014** [comment]
> PTAL @Defined2014

Please add a test case in `cmd/explaintest/t/new_character_set_builtin.test`, reset LGTM.

/cc @zimulala @xiongjiwei @tangenta 
PTAL

**Comment by hawkingrei** [comment]
> > PTAL @Defined2014
> 
> Please add a test case in `cmd/explaintest/t/new_character_set_builtin.test`, and run `./run-tests.sh -s ../../bin/tidb-server -r all` to get the result, reset LGTM.
> 
> /cc @zimulala @xiongjiwei @tangenta PTAL

just run

```
./run-tests.sh -r new_character_set_builtin
```

**Comment by zimulala** [comment]
> > PTAL @Defined2014
> 
> Please add a test case in `cmd/explaintest/t/new_character_set_builtin.test`, and run `./run-tests.sh -s ../../bin/tidb-server -r all` to get the result, reset LGTM.
> 
> /cc @zimulala @xiongjiwei @tangenta PTAL

Please address this comment.

**Review by tangenta — APPROVED** [review]

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**